x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/PersistentCache.java

```diff
@@ -0,0 +1,623 @@
+        }
+    }
+
+    public void addCacheFile(CacheFile cacheFile, SortedSet<Tuple<Long, Long>> ranges) throws IOException {
+        getWriter(cacheFile).updateCacheFile(cacheFile, ranges);
+    }
+
+    public void removeCacheFile(CacheFile cacheFile) throws IOException {
+        getWriter(cacheFile).deleteCacheFile(cacheFile);
+    }
+
+    /**
+     * This method repopulates the {@link CacheService} by looking at the files on the disk and for each file found, retrieves the latest
+     * synchronized information and puts the cache file into the searchable snapshots cache.
+     *
+     * This method iterates over all node data paths and all shard directories in order to found the "snapshot_cache" directories that
+     * contain the cache files. When such a directory is found, the method iterates over the cache files and looks up their name/UUID in
+     * the existing Lucene documents that were loaded when instanciating the persistent cache index). If no information is found (ie no
+     * matching docs in the map of Lucene documents) then the file is deleted from disk. If a doc is found the stored fields are extracted
+     * from the Lucene document and are used to rebuild the necessary {@link CacheKey}, {@link SnapshotId}, {@link IndexId}, {@link ShardId}
+     * and cache file ranges objects. The Lucene document is then indexed again in the new persistent cache index (the current
+     * {@link CacheIndexWriter}) and the cache file is added back to the searchable snapshots cache again. Note that adding cache
+     * file to the cache service might trigger evictions so previously reindexed Lucene cache files might be delete again (see
+     * CacheService#onCacheFileRemoval(CacheFile) method which calls {@link #removeCacheFile(CacheFile)}.
+     *
+     * @param cacheService the {@link CacheService} to use when repopulating {@link CacheFile}.
+     */
+    void loadCacheFiles(CacheService cacheService) {
+        ensureOpen();
+        try {
+            for (CacheIndexWriter writer : writers) {
+                final NodeEnvironment.NodePath nodePath = writer.nodePath();
+                logger.debug("loading persistent cache on data path [{}]", nodePath);
+
+                for (String indexUUID : nodeEnvironment.availableIndexFoldersForPath(nodePath)) {
+                    for (ShardId shardId : nodeEnvironment.findAllShardIds(new Index("_unknown_", indexUUID))) {
+                        final Path shardDataPath = writer.nodePath().resolve(shardId);
+                        final Path shardCachePath = getShardCachePath(new ShardPath(false, shardDataPath, shardDataPath, shardId));
+
+                        if (Files.isDirectory(shardCachePath)) {
+                            logger.trace("found snapshot cache dir at [{}], loading cache files from disk and index", shardCachePath);
+                            Files.walkFileTree(shardCachePath, new CacheFileVisitor(cacheService, writer, documents));
```

Comment:
NIT: This visitor may be easier to read if it was just inlined and one doesn't have to jump around to the other class to figure out what's going on (also saves a few lines for the constructor)?

```java
                            logger.trace("found snapshot cache dir at [{}], loading cache files from disk and index", shardCachePath);
                            Files.walkFileTree(shardCachePath, new SimpleFileVisitor<>() {
                                @Override
                                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
                                    try {
                                        final String id = buildId(file);
                                        final Document cacheDocument = documents.get(id);
                                        if (cacheDocument != null) {
                                            logger.trace("indexing cache file with id [{}] in persistent cache index", id);
                                            writer.updateCacheFile(id, cacheDocument);

                                            final CacheKey cacheKey = buildCacheKey(cacheDocument);
                                            logger.trace("adding cache file with [id={}, cache key={}]", id, cacheKey);
                                            final long fileLength = getFileLength(cacheDocument);
                                            cacheService.put(cacheKey, fileLength, file.getParent(), id, buildCacheFileRanges(cacheDocument));
                                        } else {
                                            logger.trace("deleting cache file [{}] (does not exist in persistent cache index)", file);
                                            Files.delete(file);
                                        }
                                    } catch (Exception e) {
                                        throw ExceptionsHelper.convertToRuntime(e);
                                    }
                                    return FileVisitResult.CONTINUE;
                                }
                            });
                        }
```

Reply:
Sure, I pushed ac6050f206ff4e096736ef19a2812c33a365b9b7